onnxruntime/python/tools/transformers/models/phi2/convert_to_onnx.py

```diff
@@ -0,0 +1,544 @@
+        self.phi_model(input_ids, attention_mask=attention_mask, past_key_values=past_key_values)
+
+        from torch._dynamo import config
+
+        config.capture_scalar_outputs = True
+
+        logging.info("Exporting Phi2 torch model to ONNX...")
+        torch.onnx.dynamo_export(
+            self.phi_model,
+            input_ids,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+            export_options=torch.onnx.ExportOptions(dynamic_shapes=True),
+        ).save(onnx_path)
+        onnx.checker.check_model(onnx_path)
+        onnx.shape_inference.infer_shapes_path(onnx_path)
+
+    def preprocess_onnx(self, onnx_path_in: str, onnx_path_out: str, func_name: str):
+        model = onnx.load_model(onnx_path_in, load_external_data=True)
+        function_name = None
+        for func in model.functions:
+            if func.name.endswith(func_name):
+                function_name = func.name
+                break
+        assert function_name is not None
+        model = self.unroll_function(model, function_name)
+        model = self.update_edges(model, self.phi2_edge_dict)
+        model = self.simplify_phi2_op_type(model)
+        model = self.remove_dropout_layer(model)
+        onnx.save_model(
+            model,
+            onnx_path_out,
+            save_as_external_data=True,
+            all_tensors_to_one_file=True,
+            location=onnx_path_out + ".data",
+        )
+
+    def optimize_phi2_onnx(self, onnx_path: str, onnx_path_opt: str):
+        from fusion_options import FusionOptions
+        from optimizer import optimize_model
+        import uuid
+
+        processed_onnx_path = f"{uuid.uuid1()}.onnx"
+        model = onnx.load_model(onnx_path, load_external_data=True)
+        model = self.process_graph_io(self.phi_config, model)
+        onnx.save_model(
+            model,
+            processed_onnx_path,
+            save_as_external_data=True,
+            all_tensors_to_one_file=True,
+            location=processed_onnx_path + ".data",
+        )
+
+        optimization_options = FusionOptions("phi")
+        optimizer = optimize_model(
+            processed_onnx_path,
+            model_type="phi",
+            num_heads=self.phi_config.num_attention_heads,
+            hidden_size=self.phi_config.hidden_size,
+            opt_level=0,
+            optimization_options=optimization_options,
+            only_onnxruntime=False,
+        )
+
+        self.erase_onnx_model(processed_onnx_path)
+
+        if self.precision == Precision.FLOAT32:
+            optimizer.save_model_to_file(onnx_path_opt, use_external_data_format=True)
+            return
+
+        if (
+            self.precision == Precision.FLOAT16 or self.precision == Precision.INT4
+        ) and self.attn_op_type != AttentionOpType.MultiHeadAttention:
+            node_block_list = [
+                "GroupQueryAttention_29",
```

Comment:
Can you add a comment to explain why three layers stay in fp32 and potential changes for finetuned models?